invisible_cities/reco/psf_functions.py

```diff
@@ -0,0 +1,141 @@
+def add_empty_sensors_and_normalize_q(df          : pd.DataFrame,
+                                      var         : List[str],
+                                      ranges      : List[List[float]],
+                                      database    : pd.DataFrame
+                                      ) -> pd.DataFrame :
+    """
+    Adds empty sensors to the hDST
+
+    Parameters
+    ----------
+    df  : dataframe (groupby by event and npeak to do it peak by peak)
+    var : dimensions to be considered.
+
+    Returns
+    ----------
+    df  : dst with empty sipm hits.
+    """
+    delta_x = np.diff(ranges[0])[0]/2
+    delta_y = np.diff(ranges[1])[0]/2
+
+    sel_x  = in_range(database.X, df.Xpeak.unique()[0] - delta_x,  df.Xpeak.unique()[0] + delta_x)
+    sel_y  = in_range(database.Y, df.Ypeak.unique()[0] - delta_y,  df.Ypeak.unique()[0] + delta_y)
+
+    sensors = database[sel_x & sel_y]
+
+    fill_dummy = np.zeros(len(sensors))
+    pd_dict    = {}
+
+    variables  = ['event', 'time', 'npeak']
+    variables.extend([f'{v}peak' for v in var])
+
+    for v in variables:
+        pd_dict[v] = np.full(len(sensors), df.loc[:,v].unique())
+    pd_dict['X'       ] = sensors.X
+    pd_dict['Y'       ] = sensors.Y
+    pd_dict['Z'       ] = np.full(len(sensors), df.loc[:,    'Z'].min())
+    for col in df.columns.values:
+        if col not in pd_dict.keys():
+            pd_dict[col]  = fill_dummy
+
+    df2 = pd.DataFrame(pd_dict)
+    df_out = df.merge(df2, on=list(df), how='outer')
+    df_out.drop_duplicates(subset=var, inplace=True, keep='first')
+    df_out['NormQ'] = df_out.Q/df_out.Q.sum()
+    df_out['nsipm'] = np.full(len(df_out), len(df))
+
+    return df_out
+
+
+def hdst_psf_processing(dsts        : pd.DataFrame,
+                        ranges      : List[List[float]],
+                        detector_db : str,
+                        run_number  : int
+                        ) -> pd.DataFrame :
+    """
+    Adds the necessary info to a hits DST to create the PSF, namely the relative position and the normalized Q.
+
+    Parameters
+    ----------
+    dsts        : hits (1 SiPM per hit).
+    ranges      : range of the PSF in each dimension.
+
+    Returns
+    ----------
+    hdst        : hits after processing to create PSF.
+    """
+    groupedDST = dsts.groupby(['event', 'npeak'], as_index=False)
+    sipm_db    = load_db.DataSiPM(detector_db, run_number)
+    if len(ranges) >= 3:
+        hdst          = groupedDST.apply(add_variable_weighted_mean, 'Z', 'E')
+        hdst          = hdst.groupby(['event', 'npeak'], as_index=False).apply(add_empty_sensors_and_normalize_q, ['X', 'Y', 'Z'], ranges, sipm_db).reset_index(drop=True)
+        hdst['RelZ']  = hdst.Z - hdst.Zpeak
+    else:
+        hdst          = groupedDST.apply(add_empty_sensors_and_normalize_q, ['X', 'Y'], ranges, sipm_db).reset_index(drop=True)
+        hdst['Zpeak'] = np.full (len(hdst), hdst.Z.min())
+        hdst['RelZ' ] = np.zeros(len(hdst))
```

Comment:
Because `add_variable_weighted_mean` modifies the object in place (it adds one column), `hdst` is the same object as `dsts` (needs to be checked, but it should be like this). This may be a bit misleading, but I am not sure of what can work as an alternative...

In any case try to break up those long lines so it is a bit more readable.